include/ck/tensor_operation/gpu/grid/gridwise_batched_gemm_softmax_gemm_xdl_cshuffle_v1.hpp

```diff
@@ -770,16 +787,63 @@ struct GridwiseBatchedGemmSoftmaxGemm_Xdl_CShuffle
                                                                    acc_thread_buf,
                                                                    num_k_block_main_loop);
 
-            // Acc0 elementwise Op
-#if CK_WORKAROUND_SWDEV_XXXXXX_ATTN_KERNEL_CLANG_CANNOT_SCAVENGE_REGISTER
-            static_for<0, acc_thread_buf.Size(), 1>{}(
-                [&](auto i) { acc_element_op(acc_thread_buf(i), acc_thread_buf[i]); });
-#else
-            static_for<0, acc_thread_buf.Size(), 1>{}([&](auto i) {
-                ElementOpPredicatedResetNaNToMinusInf<PadN>{}.Run(
-                    acc_thread_buf(i), acc_element_op, acc_thread_buf[i]);
-            });
-#endif
+            // do MNK padding or upper triangular masking
+            if constexpr(MaskOutUpperTriangle || PadN)
+            {
+                const index_t nstart = gemm1_k_block_outer_index * NPerBlock;
+
+                static_for<0, m0, 1>{}([&](auto m0_i) {
```

Comment:
I understand this task is urgent so there is no time for refactoring this nested loop. Do you plan to refactor the nested loop in next few PRs? Something like a visitor class below so we can support more diverse masking conditions by swapping MaskOutPredicate. We can even reuse the exact same MaskOutPredicate in the host reference code.

```c++
/* MaskOutPredicate takes m_global, n_global as input and returns bool as output */ 
template <typename MaskOutPredicate>
struct AccTileVisitor
{
  void operator()(const acc_thread_buf& acc_thread_buf, index_t m_start, index_t n_start, ...) 
  {
    { // loop over acc_offset and somehow calculate respective m_global, n_global
      if (MaskOutPredicate(m_global, n_global))
      {
        acc_thread_buf(acc_offset) = -ck::NumericLimits<float>::Infinity();
      }
    } 
  }
};
```

I can do it instead In case you have other priorities.

Reply:
The concept of `MaskOutPredicate` itself is mostly the same as to `C0MatrixMask` Chao suggested above https://github.com/ROCmSoftwarePlatform/composable_kernel/pull/425#discussion_r974676261. Whatever is convenient for us will do

Reply:
I've already used C0MatrixMask. We can do MaskOutPredictor in a further PR.